File: solutions/0347-top-k-frequent-elements/solution.cpp
#include <bits/stdc++.h>
using namespace std;

class Solution {
public:
    vector<int> topKFrequent(vector<int>& nums, int k) {
        unordered_map<int, int> freq;
        priority_queue<pair<int, int>, vector<pair<int,int>>, greater<pair<int,int>>> pq;
        pair<int, int> temp;
        vector<int> ans;
        for(const int& num: nums){
            freq[num]++;
        }
        for(const auto& val: freq){
            if((int)pq.size()<k){
                temp = {val.second, val.first};
                pq.push(temp);
            }
            else if (val.second > pq.top().first){
                temp = {val.second, val.first};
                pq.push(temp);
                pq.pop();
            }
            
            
            
        }
        for(int i = 0; i < k; i++){
            ans.push_back(pq.top().second);
            pq.pop();
        }
        return ans;
        
    }
};

int main() {
    Solution sol;
    vector<int> nums = {1, 1, 1, 2, 2, 3};
    for (int x : sol.topKFrequent(nums, 2)) cout << x << " ";
    cout << endl;
}

File: solutions/0347-top-k-frequent-elements/solution.py
from collections import defaultdict
from typing import List
class Solution:
    def topKFrequent(self, nums: List[int], k: int) -> List[int]:
        d = {}
        for n in nums:
            d[n] = d.get(n, 0) + 1
        res = sorted(d, key=lambda x:d[x], reverse=True)
        return res[:k]

class Solution:
    def topKFrequent(self, nums: List[int], k: int) -> List[int]:
        d = {}
        for n in nums:
            d[n] = d.get(n, 0) + 1

        bucket = [[] for _ in range(len(nums) + 1)]
        for num, freq in d.items():
            bucket[freq].append(num)

        sol = []
        for i in range(len(bucket) - 1, 0, -1):
            for num in bucket[i]:
                sol.append(num)
                if len(sol) == k:
                    return sol
        return sol
        
        

if __name__=="__main__":
    s = Solution()

    #print(s.topKFrequent([1,1,1,2,2,3], 2))

    s = Solution2()
    
    print(s.topKFrequent([1,1,1,2,2,3], 2))